packages/entity-server/src/models/types.ts

```diff
@@ -6,12 +6,49 @@
 import { DatabaseModel, DatabaseType } from '@tupaia/database';
 import { ObjectLikeKeys, Flatten } from '../types';
 
-type PartialOrArray<T> = {
-  [field in keyof T]?: T[field] | T[field][];
+type FilterComparators = '!=' | 'ilike';
+
+type AdvancedFilterValue<T> = {
+  comparator: FilterComparators;
+  comparisonValue: T | T[];
+};
+
+/**
+ * Example:
+ * {
+ *    type: 'facility',
+ *    name: {
+ *        comparator: 'ilike',
+ *        comparisonValue: '%Pallet%'
+ *    }
+ * }
+ */
+type NormalCriteria<T> = {
+  [field in keyof T]?: T[field] | T[field][] | AdvancedFilterValue<T[field]>;
 };
 
-export type DbConditional<T> = PartialOrArray<Omit<T, ObjectLikeKeys<T>>> &
-  PartialOrArray<Flatten<Pick<T, ObjectLikeKeys<T>>, '->>'>>;
+export enum QueryConjunctions {
+  AND = '_and_',
+  OR = '_or_',
+  RAW = '_raw_',
+}
```

Comment:
I've actually tried that before. However typescript couldn't resolve the keys coming from @tupaia/database. So that's why I sorta gave up and just created an enum here :(:

![image](https://user-images.githubusercontent.com/63621318/126748814-ba49e9eb-265a-49bb-996f-5255e1c559e7.png)

Reply:
Ah! Perhaps try `[field in keyof typeof QUERY_CONJUNCTIONS]`? I think I mistyped